### PR TITLE
FEAT-#6983: Add Pluggable Documentation Module Support

### LINKED ISSUE #6983: Documentation plugin
Right now, we always use pandas docstrings, but it would be useful to support different docstrings for different engines to indicate things like supported parameters and examples.

### PR DESCRIPTION
**Pull request by sfc-gh-dpetersohn**
## What do these changes do?

Add support for custom documentation defined at runtime.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6983 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Review by sfc-gh-dpetersohn — COMMENTED** [review]
This PR doesn't change anything about the current behavior when it comes to docs, it only applies changes when the developer wants to modify the docs with this class.

We may want to use this in the future to have different docs for Ray, Dask, or other engines.

File: modin/config/test/docs_module/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from .classes import DataFrame, Series
+
+
+def read_csv():
```
Comment (YarShev) [inline modin/config/test/docs_module/__init__.py:4]: Should we put this in functions.py or methods.py?

File: modin/config/test/docs_module/classes.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class DataFrame:
```
[the remark this answers is not in the document]
Comment (sfc-gh-dpetersohn) [reply modin/config/test/docs_module/classes.py:14]: I always forget 🤣

**Comment by YarShev** [comment]
> This PR doesn't change anything about the current behavior when it comes to docs, it only applies changes when the developer wants to modify the docs with this class.
> 
> We may want to use this in the future to have different docs for Ray, Dask, or other engines.

Could you elaborate a bit on why we may want to have different docs Ray, Dask, or other engines?

**Comment by sfc-gh-dpetersohn** [comment]
>Could you elaborate a bit on why we may want to have different docs Ray, Dask, or other engines?

Some engines may have different expectations, particularly databases. If a database engine doesn't have the same guarantees as Ray or Dask, it should be reflected in the documentation.

**Comment by YarShev** [comment]
@sfc-gh-dpetersohn, got it, thanks for the clarification. CI is still failing, please take a look. For instance, lint (pydocstyle) failed because of the following error.

```bash
D200: One-line docstring should fit on one line with quotes (found 3)
```

**Comment by sfc-gh-dpetersohn** [comment]
Thanks for the ping @YarShev, I think I fixed everything (we will see if it passes 😄). It seems we are not super consistent with `D200` on pydocstyle since there are multiple instances of the docstrings being formatted this way. I will fix the new ones in this PR but maybe we should consider more standardization in the future?

**Comment by YarShev** [comment]
> Thanks for the ping @YarShev, I think I fixed everything (we will see if it passes 😄). It seems we are not super consistent with `D200` on pydocstyle since there are multiple instances of the docstrings being formatted this way. I will fix the new ones in this PR but maybe we should consider more standardization in the future?

I see that we are really not consistent with `D200`. We just put `noqa: D200` in those places but we should definitely consider more standardization on this matter in the future.

**Comment by YarShev** [comment]
@sfc-gh-dpetersohn, `build docs` is the last job left to fix. Can you take a look? It seems the issue relates to docstrings from which we build docs.

**Comment by sfc-gh-dpetersohn** [comment]
@YarShev finally figured out the issue

**Review by YarShev — APPROVED** [review]
@sfc-gh-dpetersohn, LGTM, thanks!

**Pull request merged**